mbed_host_tests/host_tests/base_host_test.py

```diff
@@ -15,28 +15,171 @@
+            '__testcase_finish',
+            '__testcase_count',
+            '__testcase_summary',
+            '__rxd_line',
+        ]
+
+        self.__assign_default_callbacks()
+
+    def __callback_default(self, key, value, timestamp):
+        """! Default callback """
+        #self.log("CALLBACK: key=%s, value=%s, timestamp=%f"% (key, value, timestamp))
         pass
 
-    def test(self, selftest):
-        """ Blocking test execution process:
+    def __assign_default_callbacks(self):
+        """! Assigns default callback handlers """
+        for key in self.__consume_by_default:
+            self.__callbacks[key] = self.__callback_default
 
-            setup()
-            test()
-            teardown()
+    def register_callback(self, key, callback, force=False):
+        """! Register callback for a specific event (key: event name)
+            @param key String with name of the event
+            @param callback Callable which will be registstered for event "key"
+            @param force God mode
         """
+
+        # Non-string keys are not allowed
+        if type(key) is not str:
+            raise TypeError("event non-string keys are not allowed")
+
+        # And finally callback should be callable
+        if not callable(callback):
+            raise TypeError("event callback should be callable")
+
+        # Check if callback has all three required parameters (key, value, timestamp)
+        # When callback is class method should have 4 arguments (self, key, value, timestamp)
+        if ismethod(callback):
+            arg_count = callback.func_code.co_argcount
+            if arg_count != 4:
+                err_msg = "callback 'self.%s('%s', ...)' defined with %d arguments"% (callback.__name__, key, arg_count)
+                err_msg += ", should have 4 arguments: self.%s(self, key, value, timestamp)"% callback.__name__
+                raise TypeError(err_msg)
+
+        # When callback is just a function should have 3 arguments func(key, value, timestamp)
+        if isfunction(callback):
+            arg_count = callback.func_code.co_argcount
+            if arg_count != 3:
+                err_msg = "callback '%s('%s', ...)' defined with %d arguments"% (callback.__name__, key, arg_count)
+                err_msg += ", should have 3 arguments: %s(key, value, timestamp)"% callback.__name__
+                raise TypeError(err_msg)
+
+        if not force:
+            # Event starting with '__' are reserved
+            if key.startswith('__'):
+                raise ValueError("event key starting with '__' are reserved")
+
+            # We predefined few callbacks you can't use
+            if key in self.__restricted_callbacks:
+                raise ValueError("we predefined few callbacks you can't use e.g. '%s'"% key)
+
+        self.__callbacks[key] = callback
+
+    def get_callbacks(self):
+        return self.__callbacks
+
+    def setup(self):
+        pass
+
+    def result(self):
```

Comment:
```result``` returns test results. So it can't return None. Also since test result is not known as this point it should have an implementation in the derived host test. Hence it ```raise NotImplementedError```. I propose to remove it completely. see above comment.